File: JS/LongestArithmeticSubsequenceOfGivenDifference.js
/**
 * @version JAVASCRIPT ECMAScript 6
 * @author MARIO SAM <eu@mariosam.com.br>
 * @see I would love to work with you instead solving web code tests: hire me!
 */

/**
 * @param {number[]} arr
 * @param {number} difference
 * @return {number}
 */
function longestSubsequence(arr, difference) {
    const dp = new Map();
    let ans = 1;
    
    for (const num of arr) {
        dp.set(num, (dp.get(num - difference) || 0) + 1);
        ans = Math.max(ans, dp.get(num));
    }

    return ans;
}
export { longestSubsequence }

File: GO/LongestArithmeticSubsequenceOfGivenDifference.go
/**
 * @version GO 1.20.5
 * @author MARIO SAM <eu@mariosam.com.br>
 * @see I would love to work with you instead solving web code tests: hire me!
 */
package GO

func longestSubsequence(arr []int, difference int) int {
	dp, ans := make(map[int]int), 1

	for _, num := range arr {
		dp[num] = dp[num-difference] + 1
		ans = max(ans, dp[num])
	}

	return ans
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
